### ISSUE TITLE
Enumerated Types (specifically shmem_cmp_t)

### ISSUE THREAD

**Issue by jamesaross** [openshmem-org/specification#175]
Pull request #32 for point-to-point synchronization replaced the `int cmp` argument with an enumerated type `shmem_cmp_t cmp`.  An enumerated type is implementation-defined.

The [C specification](http://www.open-std.org/jtc1/sc22/wg14/www/docs/n1548.pdf), 6.7.2.2 paragraph 4 states:
> Each enumerated type shall be compatible with char, a signed integer type, or an unsigned integer type. The choice of type is implementation-defined) but shall be capable of representing the values of all the members of the enumeration.

Thus, there is potential for breaking the interface if the OpenSHMEM library and application are built with different compilers that treat enumerated types differently. The probability may be low, but we have enough problems.

In general, the use of enumerated types is not type safe nor storage class safe.  Using `uint32_t` or `uint64_t` is preferable.

**Comment by nspark** [comment]
@jdinan As our process master, if we decide we want to change the `shmem_cmp_t` enumerators, is there still wiggle room for making this change prior to the upcoming RCM?

**Comment by jamesaross** [comment]
I did some testing of this for an application and library compiled under GCC 7.1 with various permutations of the enum size using the -fshort-enums on x86_64.  GCC uses the calling convention to push 4 bytes on the stack regardless of whether the enum is 1, 2, or 4 bytes.  So it probably shouldn't be broken on the most common platform.

There's one more issue with passing values to functions in this manner.  The problem also applies to prior versions of OpenSHMEM that used `int cmp`.  If an OpenSHMEM library is compiled with one set of values and they are defined differently within the shmem.h header of another implementation used to compile the application, the result is undefined.  This could be fixed by defining specific constant values for Table 7.  Otherwise, care must be taken to use the correct header for the preferred OpenSHMEM library.

Or one could expand the API with a new family of `shmem_<TYPENAME>_wait_until_[eq|ne|gt|ge|lt|le]` and `shmem_<TYPENAME>_test_[eq|ne|gt|ge|lt|le]` routines and just fix all these issues.

**Comment by BryantLam** [comment]
>  If an OpenSHMEM library is compiled with one set of values and they are defined differently within the shmem.h header of another implementation used to compile the application, the result is undefined.

OpenSHMEM doesn't have an ABI so newer revisions of one vendor's library implementation could also use different constants vs. previous versions (but in practice, this shouldn't happen because it would cause havoc with their own users).

> This could be fixed by defining specific constant values for Table 7. 

I agree that we should define these constant values. It could fall under an ABI specification if we make it a priority in a future release.

**Comment by jdinan** [comment]
There is certainly risk of breaking API backward compatibility, e.g., the following may break because of loss of precision when converting ```int``` to ```enum```:
```
int cmp = SHMEM_CMP_EQ;
...
shmem_wait_until(&val, cmp, 0);
```
Tools may also need to support multiple type signatures for these routines, depending on whether the target OpenSHMEM library supports the 1.3 or 1.4 API.  As a workaround, an implementation may be able to force the compiler to choose ```int``` as the enumeration type by adding enumeration members with values of -1 (eliminating ```unsigned int```) and 256 (eliminating ```char```) (note, avoid including ```limits.h```).

In terms of process, there is no way to reverse this change without delaying ratification of the specification.  I believe that we can flag this as an open issue, which will allow us to apply this change and use one additional meeting to ratify the spec (e.g. in December).

**Comment by BryantLam** [comment]
My opinion is that this issue is more of an ambiguity than a defect. The spec doesn't mention the `shmem_cmp_t` type has to be a *C* enum, only that it's some ill-defined enumeration type. An implementation could simply enumerate using:
```
typedef int shmem_cmp_t;
#define SHMEM_CMP_EQ 1
#define ...
```

**Comment by nspark** [comment]
@jamesaross @BryantLam @jdinan Please review 626f452 and reply whether that commit resolves this issue for the next 1.4 RCM.

**Comment by spotluri** [comment]
Closing this. I have just merged spotluri#11 into the chapter draft.
